File: services/boardbox_catalog/src/main/java/com/demoy/boardbox_catalog/controller/UserController.java
package com.demoy.boardbox_catalog.controller;

import com.demoy.boardbox_catalog.dto.UserDto;
import com.demoy.boardbox_catalog.service.UserService;
import lombok.RequiredArgsConstructor;
import org.springframework.web.bind.annotation.*;
import reactor.core.publisher.Mono;

import java.util.UUID;

@RestController
@RequiredArgsConstructor
@RequestMapping("/api/users")
public class UserController {

    private final UserService userService;

    @PostMapping
    public Mono<UserDto> createUser(@RequestBody UserDto userDto) {
        return userService.createUser(userDto);
    }

    @GetMapping("/{id}")
    public Mono<UserDto> getUserById(@PathVariable UUID id) {
        return userService.getUserById(id);
    }

    @GetMapping("/username/{username}")
    public Mono<UserDto> getUserByUsername(@PathVariable String username) {
        return userService.getUserByUsername(username);
    }

    @PutMapping("/{id}")
    public Mono<UserDto> updateUser(@PathVariable UUID id, @RequestBody UserDto userDto) {
        return userService.updateUser(id, userDto);
    }

    @DeleteMapping("/{id}")
    public Mono<Void> deleteUser(@PathVariable UUID id) {
        return userService.deleteUser(id);
    }
}

File: services/order-service/src/main/kotlin/com/demoy/orderservice/controller/UserController.kt
package com.demoy.orderservice.controller

import com.demoy.orderservice.model.AppUser
import com.demoy.orderservice.service.UserService
import org.springframework.web.bind.annotation.*
import reactor.core.publisher.Mono
import java.time.LocalDateTime
import java.util.*

data class CreateUserRequest(
    val name: String,
    val email: String? = null
)

data class UserResponse(
    val id: String,
    val name: String,
    val email: String?,
    val createdAt: LocalDateTime
)

@RestController
@RequestMapping("/api/users")
class UserController(private val userService: UserService) {

    @PostMapping
    fun create(@RequestBody req: CreateUserRequest): Mono<UserResponse> {
        return userService.createUser(req.name, req.email)
            .map { it.toResponse() }
    }

    @GetMapping("/{userId}")
    fun get(@PathVariable userId: String): Mono<UserResponse> {
        val uid = UUID.fromString(userId)
        return userService.getUser(uid).map { it.toResponse() }
    }

    @GetMapping(params = ["email"])
    fun getByEmail(@RequestParam email: String): Mono<UserResponse> {
        return userService.getUserByEmail(email).map { it.toResponse() }
    }

    private fun AppUser.toResponse(): UserResponse =
        UserResponse(id = this.id.toString(), name = this.name, email = this.email, createdAt = this.createdAt)
}
